plugin-build/src/main/kotlin/io/sentry/android/gradle/telemetry/SentryTelemetryService.kt

```diff
@@ -0,0 +1,474 @@
+            } ?: false
+        return isSentryTaskName ||
+            throwable.stackTrace.any { it.className.startsWith("io.sentry") }
+    }
+
+    fun captureError(exception: Throwable, operation: String?) {
+        val message = if (exception is SentryCliException) {
+            "$operation failed with SentryCliException and reason ${exception.reason}"
+        } else {
+            "$operation failed with ${exception.javaClass}"
+        }
+
+        val mechanism = Mechanism().also {
+            it.type = MECHANISM_TYPE
+            it.isHandled = false
+        }
+        val mechanismException: Throwable =
+            ExceptionMechanismException(
+                mechanism,
+                SentryMinimalException(message),
+                Thread.currentThread()
+            )
+        val event = SentryEvent(mechanismException).also {
+            it.level = SentryLevel.FATAL
+        }
+        hub.captureEvent(event)
+    }
+
+    fun startRun(transactionName: String) {
+        hub.startSession()
+        transaction = hub.startTransaction(transactionName, "build", true)
+    }
+
+    fun endRun() {
+        if (didAddChildSpans) {
+            transaction?.finish()
+            hub.endSession()
+        }
+    }
+
+    fun traceCli(): List<String> {
+        val args = mutableListOf<String>()
+        hub.traceparent?.let { header ->
+            args.add("--header")
+            args.add("${header.name}:${header.value}")
+        }
+        hub.baggage?.let { header ->
+            args.add("--header")
+            args.add("${header.name}:${header.value}")
+        }
+        return args
+    }
+
+    fun startTask(operation: String): ISpan? {
+        didAddChildSpans = true
+        hub.setTag("step", operation)
+        return hub.span?.startChild(operation)
+    }
+
+    fun endTask(span: ISpan?, task: Task) {
+        span?.let { span ->
+            task.state.failure?.let { throwable ->
+                captureError(throwable, span.operation)
+                span.status = SpanStatus.UNKNOWN_ERROR
+            }
+
+            span.finish()
+        }
+    }
+
+    override fun close() {
+        if (transaction?.isFinished == false) {
+            endRun()
+        }
+        Sentry.close()
+    }
+
+    companion object {
+        val SENTRY_SAAS_DSN: String =
+            "https://dd1f82ad30a331bd7def2a0dce926c6e@o447951.ingest.sentry.io/4506031723446272"
+        val MECHANISM_TYPE: String = "GradleTelemetry"
+        private val orgRegex = Regex("""(?m)Default Organization: (.*)$""")
+
+        fun createParameters(
+            project: Project,
+            variant: SentryVariant?,
+            extension: SentryPluginExtension,
+            cliExecutable: String,
+            sentryOrg: String?,
+            buildType: String
+        ): SentryTelemetryServiceParams {
+            val tags = extraTagsFromExtension(project, extension)
+            var isSaas: Boolean? = null
```

Comment:
`m`: I guess we should check here for `extension.url.orNull != null` or something like that?